ARO-8948 Update ARO HCP template with CSI Driver

ahitacat opened the pull request:
### What this PR does

This PR adds a SecretClassProvider that will store the first-party credential for CS deployment.
Additionally, a new make target and documentation have been added.

Jira:[ ARO-8948](https://issues.redhat.com/browse/ARO-8948)
Link to demo recording: <!-- optional: link to a demo recording -->

### Special notes for your reviewer

<!-- optional -->

ahitacat commented:
@microsoft-github-policy-service agree company="Red Hat"

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
 - name: CS_SERVICE_PRINCIPAL_CREDS_BASE64
   description: Base64 encoded service principal credentials for CS. This is temporary for P1
   required: true
+- name: KEYVAULT_NAME
```
@bennerv will there be a case where these types of credentials/certificates are moved to their own key vault? 

I'd like to make the name of this key vault specific. For example we also have the arm helper identity credentials. Are these going to be stored in the same key vault? cc @miguelsorianod 

Maybe something like `AZURE_FIRST_PARTY_APPLICATION_CREDENTIALS_KEYVAULT_NAME` or `AZURE_IDENTITY_CREDENTIALS_KEYVAULT_NAME`?

ahitacat replied:
Would we like to create different key vaults to store different secrets? 
If so I'm ok with changing it, but this PR stores it in the private service-kv. Thus, I will propose (at this moment) to change it to `SERVICE_KEYVAULT_NAME`. What do you think?

JameelB replied:
I think we should make the name meaningful to CS as the current kv name may change. It may also not have the same name for other environments. I think `SERVICE_KEYVAULT_NAME` is a bit ambiguous though, what does `SERVICE` mean? Is it related to cluster service, the service cluster, or something else? 🤔 

As for multiple kv vs one kv, I think @bennerv or @janboll is best to answer that 🙏

ahitacat replied:
Make sense, I will wait for the answer of multiple kv vs one, so we can select one of your suggestions. As one describes a general kv than the other.

bennerv replied:
service comes from the name of the key vault used to support the service running the resource provider.  There will be other secrets in the service key vault which contain TLS certificates, MDSD certificates (potentially) etc.  

It might make sense to align on naming across the components so we're all talking about the same thing, but it's up to CS to determine what they want to do with the name.

ahitacat replied:
Then I'll opt for AZURE_IDENTITY_CREDENTIALS_KEYVAULT_NAME as it seems it will potentially contain other certs.

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
   description: Base64 encoded service principal credentials for CS. This is temporary for P1
   required: true
+- name: KEYVAULT_NAME
+  description: The name of the Key Vault where the first party credentials are stored.
```
```suggestion
  description: The name of the Key Vault where the first party application credentials are stored.
```

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+- name: KEYVAULT_NAME
+  description: The name of the Key Vault where the first party credentials are stored.
+  value: "service-kv-aro-hcp-dev"
+- name: TENANT_ID
```
Let's specify that this is the keyvault tenant id  in the name. 

Similar to my comment above, are all of our key vaults going to live in the same tenant? Or should we make this specific per environment. Maybe something like `AZURE_FIRST_PARTY_APPLICATION_KEYVAULT_TENANT_ID` or `AZURE_IDENTITY_CREDENTIALS_KEYVAULT_TENANT_ID` (where the arm helper identities can also use this tenant id)?

ahitacat replied:
Similar question from above. In addition, this tenant is the same as the one required for maestro and hypershift. Note that for maestro we create a new key vault to store its related secrets.

JameelB replied:
Note that we also provide other tenant ids for other purposes. For example, in the current aro auth config, there is a tenant id there used for authenticating clients within CS. I think it should be clear in the name as to what the tenant will be used for to avoid confusion and prevent people using this for different purposes other than what it's intended for.

bennerv replied:
TenantID should always be the same no matter what key vault you're accessing.  The only difference would be if you're using the 1p credential to auth into the customer's tenant, but that's not a configuration that's set as a runtime env var or command-line flag.

JameelB replied:
Then in that case, maybe just `KEYVAULT_TENANT_ID` would suffice? or to keep it in line with the service keyvault, `SERVICE_KEYVAULT_TENANT_ID`? 

There is another TenantID that we'd need to set as an env var, its the tenant id of where CS is running on and will be added/documented as part of https://issues.redhat.com/browse/ARO-7425

Will the tenant where CS is running on and where the service keyvault is located always be the same?

bennerv replied:
Tenant will always be the same in which the service keyvault exists and CS runs.  No need to make it specific to the key vault instance.

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+  value: "service-kv-aro-hcp-dev"
+- name: TENANT_ID
+  description: The Tenant ID of Key Vault.
+  value: ""
```
Since we have a default value for the keyvault name, I think we should then specify the tenant id of that keyvault here as default value.

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+- apiVersion: secrets-store.csi.x-k8s.io/v1
+  kind: SecretProviderClass
+  metadata:
+    name: cs-keyvault
```
name may need to change based on how we're organising keyvault identities. It may be the case we'll need multiple SecretProviderClass if we have multiple keyvaults to integrate with cs.

ahitacat replied:
Related to the same questions above.

bennerv replied:
We'll have multiple key vaults, but I believe this is the only one which will be used to mount a secret in the cluster for CS usage.  The others CS will access the data plane of the key vaults directly, rather than leveraging secrets CSI driver.

JameelB replied:
i'm fine with leaving it with this name then, feel free to close thread as resolved 👍

ahitacat replied:
I can use something like `cs-azure-id-credentials-keyvault` to specify what the keyvault is for.

JameelB reviewed (commented):
Thanks for this @ahitacat! lgtm overall, some comments about naming. Maybe @bennerv, @miguelsorianod or @janboll can help here to give context on how these identity credentials may be organized.

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -465,6 +498,9 @@ objects:
+          - name: keyvault
+            mountPath: "/secrets/keyvault"
+            readOnly: true
           env:
```
Can we also add the flag `azure-first-party-application-certificate-bundle-path` since the file path is now known? 🙏  You may need to update the image tag of cs to an image tag that has this change, similar to what's done in this mr: https://github.com/Azure/ARO-HCP/pull/510/files

The latest tag in that repo `app-sre/uhc-clusters-service` is currently `1c6873c`, which matches the latest commit on cs master. Any reason not to use the latest tag @miguelsorianod @bennerv @janboll ?

bennerv replied:
It looks added here.  I don't see a reason not to use it if CS has functionality for it.

ahitacat commented:
## Latest Changes:

- Added a new commit that adds the `azure-first-party-application-certificate-bundle-path` and updates clusters-service tag to the latest.
- Added Tenant ID default value, and fixed a minor description.

I'm waiting for the names of the Keyvault and Tenant ID parameters questions to be resolved.
This PR will need to be closed and I need to make the changes in a branch of this project to make the pipelines run. But I prefer to do this once the review has finished.

bennerv commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+    namespace: ${NAMESPACE}
+  spec:
+    parameters:
+      clientID: ${AZURE_CS_MI_CLIENT_ID}
```
@geoberle should this be CS' identity used here? Shouldn't we have a dedicated identity for secrets CSI storage driver to mount the secret into the pod, rather than give it broad permissions CS has?

geoberle replied:
good point. i would not hold off this PR from being merged because of this though. this is something we should do for other services as well, e.g. maestro-server

i've created some cards to track this work
https://issues.redhat.com/browse/ARO-10155

ahitacat replied:
Once those identities are created they can be changed here. Resolving the conversation.

geoberle commented on cluster-service/Makefile:
```diff
@@ -20,13 +21,16 @@ deploy:
 	  -p AZURE_CS_MI_CLIENT_ID=$${AZURE_CS_MI_CLIENT_ID} \
+	  -p TENANT_ID=$${TENANT_ID} \
 	  -p REGION=${REGION} \
+	  -p KEYVAULT_NAME=${KEYVAULT_NAME} \
```
should this be named `SERVICE_KEYVAULT_NAME` to match the parameter name expected by the template?

ahitacat replied:
good catch

geoberle commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+  value: "service-kv-aro-hcp-dev"
+- name: TENANT_ID
+  description: The Tenant ID of Key Vault.
+  value: "64dc69e4-d083-49fc-9569-ebece1dd1408"
```
i would prefer to not provide a default here targeting a specific tenant

ahitacat replied:
I would prefer that too, @JameelB wdyt?

JameelB commented on cluster-service/deploy/openshift-templates/arohcp-service-template.yml:
```diff
@@ -222,8 +222,35 @@ parameters:
+      objects: |-
+        array:
+          - |
+            objectName: firstPartyMock
+            objectType: secret
+            objectAlias: "firstPartyMock"
```
Been thinking about this, is it possible to parameterise it like the keyvault name? I assume for the different environments, the name isn't going to have 'mock' on them 🤔 wdyt?

ahitacat commented:
Superseded by https://github.com/Azure/ARO-HCP/pull/597
